### PR TITLE
🌱 Add conformance and upgrade e2e tests

### LINKED ISSUE #2183: Consider introducing a conformance test which tests against Kubernetes ci latest
The high-level idea is to introduce a separate conformance test job which tests against Kubernetes latest version (e.g. after the 1.28 release against 1.29)

This would work similarly to other infra providers like CAPZ / CAPO. Basically during the e2e test we would install the latest ci binaries on Machine bootstrap. 

Notes:
* There is a GenerateCIArtifactsInjectedTemplateForDebian util func in core CAPI which is usually used to inject the installation script to install ci latest binaries / controller images / ...

P.S. For now I would consider this low priority relative to other things we're working on

### PR DESCRIPTION
**Pull request by sbueringer**
Signed-off-by: Stefan Büringer buringerst@vmware.com

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2183
Part of https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/issues/2666

### CODE REVIEW

File: test/e2e/data/infrastructure-vsphere/main/clusterclass/patch-vsphere-template.yaml [hunk 1]
```diff
@@ -6,11 +6,15 @@
       - op: replace
         path: /spec/template/spec/template
         valueFrom:
+          # We currently only use v1.28.0 & v1.29.0 images with this ClusterClass which both
```
Comment (sbueringer) [inline test/e2e/data/infrastructure-vsphere/main/clusterclass/patch-vsphere-template.yaml:9]: TODO: Also has to still work with 1.27.3 for clusterctl upgrade tests

**Comment by sbueringer** [comment]
/hold

for corresponding test-infra PR

**Comment by sbueringer** [comment]
/test pull-cluster-api-provider-vsphere-conformance-ci-latest-main
/test pull-cluster-api-provider-vsphere-conformance-main

(upgrade tests are not available yet before the test-infra PR is merged)

**Review by chrischdi — COMMENTED** [review]
/lgtm

Should we wait for test-infra to merge so we can run the tests?

**Comment by sbueringer** [comment]
> /lgtm
> 
> Should we wait for test-infra to merge so we can run the tests?

I'm fine either way

**Comment by chrischdi** [comment]
> @sbueringer: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> 
> Test name	Commit	Details	Required	Rerun command
> pull-cluster-api-provider-vsphere-e2e-conformance-ci-latest-main	[bfd9cb6](https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/commit/bfd9cb622139a00685d65f1455f2c9c5009a5418)	[link](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-vsphere/2594/pull-cluster-api-provider-vsphere-e2e-conformance-ci-latest-main/1750928106421489664)	true	`/test pull-cluster-api-provider-vsphere-e2e-conformance-ci-latest-main`
> [Full PR test history](https://prow.k8s.io/pr-history?org=kubernetes-sigs&repo=cluster-api-provider-vsphere&pr=2594). [Your PR dashboard](https://prow.k8s.io/pr?query=is%3Apr%20state%3Aopen%20author%3Asbueringer). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc) when you hit one in your PR.

> WARNING: The requested image's platform (linux/arm64/v8) does not match the detected host platform (linux/amd64/v3) and no specific platform was requested

**Comment by chrischdi** [comment]
I think we need this resolve version logic:

https://github.com/kubernetes-sigs/cluster-api/blob/419775cd357377c8aca8b59250afee0132abbac2/scripts/ci-e2e-lib.sh#L62-L63

to fix https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-vsphere/2594/pull-cluster-api-provider-vsphere-e2e-upgrade-1-29-1-30-main/1750934298506039296

**Comment by sbueringer** [comment]
Thx, I think you're right. Pushed a fix.

/test pull-cluster-api-provider-vsphere-e2e-conformance-ci-latest-main
/test pull-cluster-api-provider-vsphere-e2e-conformance-main
/test pull-cluster-api-provider-vsphere-e2e-main
/test pull-cluster-api-provider-vsphere-e2e-upgrade-1-29-1-30-main

**Comment by sbueringer** [comment]
@chrischdi I think some trouble with IPAM here: pull-cluster-api-provider-vsphere-e2e-main

**Comment by sbueringer** [comment]
@chrischdi wdyt about merging this PR anyway and then monitoring the situation in test grid?

Looks like some comformance tests are failing.

**Comment by sbueringer** [comment]
Hm this was testing 1.28 => 1.29 upgrade because I red the variables from the e2e config directly instead of using GetVariable which also considers env vars. Let's give this another try

(But I guess 1.28 =>. 1.29 should have also worked..)

**Comment by chrischdi** [comment]
https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/pull/2678#event-11646683779 should help for the next time the issue of https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-vsphere/2594/pull-cluster-api-provider-vsphere-e2e-upgrade-1-29-1-30-main/1752333834268971008 appears.


/test pull-cluster-api-provider-vsphere-e2e-upgrade-1-29-1-30-main

**Comment by chrischdi** [comment]
> @chrischdi wdyt about merging this PR anyway and then monitoring the situation in test grid?
> 
> Looks like some comformance tests are failing.

I'm okay with that.

Latest failure does not point to a conformance test failure on a first view.

/lgtm

/hold

feel free to unhold

**Comment by chrischdi** [comment]
```
[2024-01-30 20:19:57] 	[ERROR ImagePull]: failed to pull image registry.k8s.io/kube-apiserver:v1.30.0-alpha.0.906_922fbbf5b70d46: output: E0130 20:19:54.923476     995 remote_image.go:167] "PullImage from image service failed" err="rpc error: code = NotFound desc = failed to pull and unpack image \"registry.k8s.io/kube-apiserver:v1.30.0-alpha.0.906_922fbbf5b70d46\": failed to resolve reference \"registry.k8s.io/kube-apiserver:v1.30.0-alpha.0.906_922fbbf5b70d46\": registry.k8s.io/kube-apiserver:v1.30.0-alpha.0.906_922fbbf5b70d46: not found" image="registry.k8s.io/kube-apiserver:v1.30.0-alpha.0.906_922fbbf5b70d46"
```

Image pull issues

**Comment by sbueringer** [comment]
> ```
> [2024-01-30 20:19:57] 	[ERROR ImagePull]: failed to pull image registry.k8s.io/kube-apiserver:v1.30.0-alpha.0.906_922fbbf5b70d46: output: E0130 20:19:54.923476     995 remote_image.go:167] "PullImage from image service failed" err="rpc error: code = NotFound desc = failed to pull and unpack image \"registry.k8s.io/kube-apiserver:v1.30.0-alpha.0.906_922fbbf5b70d46\": failed to resolve reference \"registry.k8s.io/kube-apiserver:v1.30.0-alpha.0.906_922fbbf5b70d46\": registry.k8s.io/kube-apiserver:v1.30.0-alpha.0.906_922fbbf5b70d46: not found" image="registry.k8s.io/kube-apiserver:v1.30.0-alpha.0.906_922fbbf5b70d46"
> ```
> 
> Image pull issues

Yes and no. Thx for the cloud-init PR, made it a lot easier to debug.

> [2024-01-30 20:19:53] 	[WARNING KubernetesVersion]: Kubernetes version is greater than kubeadm version. Please consider to upgrade kubeadm. Kubernetes version: 1.30.0-alpha.0.906+922fbbf5b70d46. Kubeadm version: 1.29.x

The problem was that I didn't use the new flavor for the upgrade test. So it tried to install 1.30 ci latest on an image that only had 1.29 Kubernetes. Should be fixed now

**Comment by sbueringer** [comment]
@chrischdi I think now we're good to go

**Comment by chrischdi** [comment]
/lgtm

/hold cancel

(note: failures are conformance test failures, not the CAPI ones)

**Comment by chrischdi** [comment]
@sbueringer so we should overwrite the failing test I guess? Should not impact dev that much because that job only runs when we trigger it right?

**Comment by sbueringer** [comment]
/override pull-cluster-api-provider-vsphere-e2e-upgrade-1-29-1-30-main

**Pull request merged**